cmd/snap/cmd_run.go

```diff
@@ -133,15 +135,32 @@ func createOrUpdateUserDataSymlink(info *snap.Info, usr *user.User) error {
 			// TRANSLATORS: %v the error message
 			return fmt.Errorf(i18n.G("cannot read symlink: %v"), err)
 		}
+
 		if currentSymlinkValue == wantedSymlinkValue {
 			break
 		}
+
 		if !os.IsNotExist(err) {
-			_ = os.Remove(currentActiveSymlink)
+			// We may be racing with other instances of snap-run that try to do the same thing
+			// If the symlink is already removed then we can ignore this error and try again.
+			err = os.Remove(currentActiveSymlink)
+			if os.IsNotExist(err) {
```

Comment:
Logic here is bogus. It's continuing when it should fallthrough (file doesn't exist), and it's breaking when it should continue (err == nil).

This should be:
```
err = os.Remove(...)
if err != nil && !os.IsNotExist(err) {
        break
}
```

That said, how can we possibly get here at all?  Right after reading the link, we have a test saying
```
if err != nil && !os.IsNotExist(err) {
 			return fmt.Errorf(i18n.G("cannot read symlink: %v"), err
}
```
So if `!os.IsNotExist(err)`, we'd never even try to remove. This logic is almost certainly untested.

The test in line 143 should probably read `if err == nil` instead, and we need a test for this.